File: 697. Degree of an Array.java
class Solution {
    public int findShortestSubArray(int[] nums) {
        HashMap<Integer, Integer> left = new HashMap<>(), right = new HashMap<>(), count = new HashMap<>();
        for (int i=0; i<nums.length; i++) {
            count.put(nums[i], count.getOrDefault(nums[i], 0)+1);
            right.put(nums[i], i);
            if (!left.containsKey(nums[i])) {
                left.put(nums[i], i);
            }
        }
        int degree = Collections.max(count.values());
        int ans = nums.length;
        for (int x : count.keySet()) {
            if (count.get(x) == degree) {
                ans = Math.min(ans, right.get(x) - left.get(x)+1);
            }
        }
        return ans;
    }
}

File: 697. Degree of an Array.py
class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        dic = collections.defaultdict(list)
        for i, n in enumerate(nums):
            dic[n].append(i)
        ls = sorted(dic.values(), key=len, reverse=True)
        deg = ls[0][-1] - ls[0][0] + 1
        for i in range(1,len(ls)):
            if len(ls[i]) < len(ls[i-1]):
                break
            deg = min(deg, ls[i][-1] - ls[i][0] + 1)
        return deg
        